firmware/eosio_token.c

```diff
@@ -0,0 +1,94 @@
+#include "eosio_token.h"
+
+#include "eos.h"
+#include "gettext.h"
+#include "protect.h"
+
+#include "messages-eos.pb.h"
+
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stdio.h>
+
+#define CHECK_COMMON(ACTION) \
+    do { \
+        CHECK_PARAM_RET(common->account == EOS_eosio_token, \
+                        _("Incorrect account name"), false); \
+        CHECK_PARAM_RET(common->name == (ACTION), \
+                        _("Incorrect action name"), false); \
+    } while(0)
+
+bool eos_compileActionTransfer(const EosActionCommon *common,
+                               const EosActionTransfer *action) {
+    CHECK_COMMON(EOS_Transfer);
+
+    CHECK_PARAM_RET(action->has_quantity, _("Required field missing"), false);
+    CHECK_PARAM_RET(action->has_sender, _("Required field missing"), false);
+    CHECK_PARAM_RET(action->has_receiver, _("Required field missing"), false);
+    CHECK_PARAM_RET(action->has_memo, _("Required field missing"), false);
+
+    size_t memo_len = strlen(action->memo);
+
+    if (256 < memo_len) {
+        fsm_sendFailure(FailureType_Failure_DataError, _("Memo too long"));
+        eos_signingAbort();
+        layoutHome();
+        return false;
+    }
+
+    char asset[EOS_ASSET_STR_SIZE];
+    CHECK_PARAM_RET(eos_formatAsset(&action->quantity, asset),
+                    _("Invalid asset format"), false);
+
+    char sender[EOS_NAME_STR_SIZE];
+    CHECK_PARAM_RET(eos_formatName(action->sender, sender),
+                    _("Invalid name"), false);
+
+    char receiver[EOS_NAME_STR_SIZE];
+    CHECK_PARAM_RET(eos_formatName(action->receiver, receiver),
+                    _("Invalid name"), false);
+
+    layoutDialogSwipe(
+        &bmp_icon_question,
+        _("Cancel"),
+        _("Confirm"),
+        _("Transfer"),
+        _("Do you want to send"),
+        asset,
+        _("from"),
+        sender,
+        _("to"),
+        receiver);
+    if (!protectButton(ButtonRequestType_ButtonRequest_ConfirmOutput, false)) {
```

Comment:
In our port, I called this button request `ConfirmAction`. What would you prefer here? It seems unfriendly to clients to call everything `ConfirmOutput`...